test/cpp/naming/naming_end2end_test.cc

```diff
@@ -0,0 +1,297 @@
+    char *str;
+    grpc_sockaddr_to_string(&str, &addr.address, 1 /* normalize */);
+    gpr_log(GPR_INFO, "%s", str);
+    GPR_ASSERT(addr.is_balancer == args->expect_is_balancer);
+    GPR_ASSERT(matches_any(args->expected_addrs, str));
+    gpr_free(str);
+  }
+
+  check_service_config_result_locked(channel_args, args);
+
+  gpr_atm_rel_store(&args->done_atm, 1);
+  gpr_mu_lock(args->mu);
+  GRPC_LOG_IF_ERROR("pollset_kick", grpc_pollset_kick(args->pollset, NULL));
+  gpr_mu_unlock(args->mu);
+}
+
+static void test_resolves(grpc_exec_ctx *exec_ctx, args_struct *args) {
+  char *whole_uri = NULL;
+  const char *authority = FLAGS_local_dns_server_address.c_str();
+  if (authority != NULL && strlen(authority) > 0) {
+    gpr_log(GPR_INFO, "Specifying authority in uris to: %s", authority);
+  } else {
+    authority = "";
+  }
+
+  GPR_ASSERT(asprintf(&whole_uri, "dns://%s/%s", authority,
+                      (char *)args->target_name));
+  args->channel_args = NULL;
+
+  grpc_resolver *resolver = grpc_resolver_create(exec_ctx, whole_uri, NULL,
+                                                 args->pollset_set, args->lock);
+  gpr_free(whole_uri);
+
+  grpc_closure on_resolver_result_changed;
+  GRPC_CLOSURE_INIT(&on_resolver_result_changed,
+                    check_channel_arg_srv_result_locked, (void *)args,
+                    grpc_combiner_scheduler(args->lock));
+
+  grpc_resolver_next_locked(exec_ctx, resolver, &args->channel_args,
+                            &on_resolver_result_changed);
+
+  grpc_exec_ctx_flush(exec_ctx);
+  poll_pollset_until_request_done(args);
+  GRPC_RESOLVER_UNREF(exec_ctx, resolver, NULL);
+}
+
+void naming_end2end_test_resolution(const char *name,
+                                    std::vector<std::string> expected_addrs,
+                                    const char *expected_service_config,
+                                    bool expected_target_is_balancer) {
+  grpc_init();
+  gpr_log(GPR_INFO, "e a size: %d", (int)expected_addrs.size());
+  grpc_exec_ctx exec_ctx = GRPC_EXEC_CTX_INIT;
+  args_struct args;
+  args_init(&exec_ctx, &args);
+  args.expect_is_balancer = expected_target_is_balancer;
+  args.target_name = name;
+  args.expected_addrs = expected_addrs;
+  args.expected_service_config_string = expected_service_config;
+
+  test_resolves(&exec_ctx, &args);
+  args_finish(&exec_ctx, &args);
+  grpc_exec_ctx_finish(&exec_ctx);
+  grpc_shutdown();
+}
+
+}  // namespace
+}  // namespace grpc
+
+int main(int argc, char **argv) {
+  grpc::testing::InitTest(&argc, &argv, true);
+  if (FLAGS_target_name == "" || FLAGS_expected_addrs == "") {
+    gpr_log(GPR_ERROR,
+            "Missing target_name or expected_addrs params. Got %s and %s",
+            FLAGS_target_name.c_str(), FLAGS_expected_addrs.c_str());
+    abort();
+  }
+  auto expected_addrs_list =
+      grpc::testing::parse_expected_addrs(FLAGS_expected_addrs.c_str());
+
+  grpc::testing::naming_end2end_test_resolution(
```

Comment:
Renamed this to `resolver_component_test..`. 

I wanted make it more unique than `resolver_test`, for namespacing/regex matching purposes, but can change if we want to